File: golang/aoc/aoc.go
package aoc

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type AocInput struct {
	Text  string
	Lines []string
}

func openFile(filePath string) (*os.File, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("unable to open file: %v", err)
	}
	return file, nil
}

func readFileLines(file *os.File) ([]string, string, error) {
	var lines []string
	var sb strings.Builder
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
		sb.WriteString(line + "\n")
	}

	if err := scanner.Err(); err != nil {
		return nil, "", fmt.Errorf("unable to read file: %v", err)
	}

	return lines, sb.String(), nil
}

func FindAocInput(filePath string) (*AocInput, error) {
	file, err := openFile(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	lines, text, err := readFileLines(file)
	if err != nil {
		return nil, err
	}

	return &AocInput{
		Text:  text,
		Lines: lines,
	}, nil
}

File: typescript/aoc.ts
import * as fs from "fs/promises";
import * as path from "path";

export interface AocInput {
    text: string;
    lines: string[];
}

export const findAocInput = async (filePath: string): Promise<AocInput> => {
    const fullPath = path.resolve(filePath);
    try {
        const text = await fs.readFile(fullPath, "utf-8");
        let lines = text.split("\n").map(line => line.replace(/\r/g, ""));
        if (lines[lines.length - 1] === "") {
            lines = lines.slice(0, -1);
        }
        return <AocInput>{ text, lines };
    } catch (error) {
        console.error(`Unable to read file at ${filePath}: ${error}`);
        throw new Error();
    }
};
